tools/data_source/upload.py

```diff
@@ -77,14 +77,28 @@ def add_file(dataset, registry, json_file, output_path):
     converted_path = None
     stdout = None
     link_data_only = dataset.get('link_data_only', 'copy_files')
-    run_as_real_user = in_place = dataset.get('in_place', True)
-    purge_source = dataset.get('purge_source', True)
-    # in_place is True if there is no external chmod in place,
-    # however there are other instances where modifications should not occur in_place:
-    # when a file is added from a directory on the local file system (ftp import folder or any other path).
-    if dataset.type in ('server_dir', 'path_paste', 'ftp_import'):
-        in_place = False
+
+    # run_as_real_user is estimated from galaxy config (external chmod indicated of inputs executed)
+    # If this is True we always purge supplied upload inputs so they are cleaned up and we reuse their
+    # paths during data conversions since this user already owns that path.
+    run_as_real_user = dataset.get('run_as_real_user', False)
+
+    # purge_source is False if this is an FTP import and ftp_upload_purge has been overridden to False in Galaxy's config.
+    # This prevents us from deleting the user supplied paths in this case. We disable this behavior
+    # if running as the real user so the file can be cleaned up by Galaxy.
+    purge_source = dataset.get('purge_source', True) and not run_as_real_user
+
+    # in_place is True only if we are running as a real user and not importing external paths (i.e.
+    # this is a real upload and not a path paste or ftp import).
+    # In this case we try to reuse the uploaded file that has been chowned to this user already.
+    in_place = run_as_real_user and dataset.type not in ('server_dir', 'path_paste', 'ftp_import')
+
+    # check_content is a flag in Galaxy's config that can be swapped off to disable checking for HTML content
+    # and such that can prevent uploads from working in some cases.
     check_content = dataset.get('check_content' , True)
+
+    # auto_decompress is a request flag that can be swapped on to prevent Galaxy from automatically
```

Comment:
s/on/off/ ?

Reply:
Yeah - good point. I'll rebase with this fix.